src/ring_theory/adjoin_root.lean

```diff
@@ -95,16 +95,56 @@ principal_ideal_domain.is_maximal_of_irreducible ‹irreducible f›
 noncomputable instance field : discrete_field (adjoin_root f) :=
 ideal.quotient.field (span {f} : ideal (polynomial α))
 
-instance : is_field_hom (coe : α → adjoin_root f) := by apply_instance
+instance : is_field_hom (of : α → adjoin_root f) := by apply_instance
+
+lemma of_injective : function.injective (@of α _ f) := is_field_hom.injective _
 
 instance lift_is_field_hom [field β] {i : α → β} [is_ring_hom i] {a : β}
   {h : f.eval₂ i a = 0} : is_field_hom (lift i a h) := by apply_instance
 
-lemma coe_injective : function.injective (coe : α → adjoin_root f) :=
-is_field_hom.injective _
-
 lemma mul_div_root_cancel (f : polynomial α) [irreducible f] :
-  (X - C (root : adjoin_root f)) * (f.map coe / (X - C root)) = f.map coe :=
+  (X - C (root : adjoin_root f)) * (f.map of / (X - C root)) = f.map of :=
 mul_div_eq_iff_is_root.2 $ is_root_root _
 
+end field
+
+section algebra
+
+variables (R : Type*) [comm_ring R] [comm_ring α] [algebra R α] (f : polynomial α) [comm_ring β]
+  [algebra R β]
+
+instance : algebra R (adjoin_root f) :=
+algebra.of_ring_hom (of ∘ algebra_map _) (is_ring_hom.comp _ _)
+
+def aof : α →ₐ[R] adjoin_root f :=
+{ commutes' := λ _, rfl, ..ring_hom.of of }
+
+lemma aof_eq_of : ⇑(@aof α R _ _ _ f) = of := rfl
+
+def amk : polynomial α →ₐ[R] adjoin_root f :=
```

Comment:
We might need to think a bit more about how to handle "large" amounts of algebras. Because `adjoin_root f` is an algebra over `polynomial α` which is an algebra over `α` which is an algebra over `R`. (Lean is not aware of all of these.) The fact that `amk` is an algebra hom over `R` is just "forgetting" that it is an algebra hom over `polynomial α`. At the moment we don't have a good mechanism for "forgetting scalars" in an algebra hom. I have no idea how to handle this in a elegant manner. @rwbarton Do you have any ideas? (Maybe we should take this discussion to Zulip.)

Reply:
Having looked at comap, I think it's best to adopt the convention of defining an alg_hom to be over the biggest sensible ring, and then use `comap` the rest of the time. I guess this is what Kenny intended.